### PR TITLE
UI support for real-ranged attacks

### PR DESCRIPTION
**Pull request by hladkyjakub**
This makes units with weapons with custom min/max_range capable of attack when controlled by player.
(I have the AI code too, but not in this PR)
For proper working requires these PR merged:
Propper max/min_range: https://github.com/wesnoth/wesnoth/pull/9203
Propper UI preview: https://github.com/wesnoth/wesnoth/pull/9255

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
For me, "multihex attacks" sounds like "an attack that hits more than one hex", which is totally different from what this PR actually is.

It would be helpful to see screenshots of this PR in action.

**Comment by hladkyjakub** [comment]
> It would be helpful to see screenshots of this PR in action.

Sure:
Screenshot_20240902_163005
Screenshot_20240902_163021
Screenshot_20240902_163128

File: src/mouse_events.cpp [hunk 1]
```diff
@@ -1484,8 +1515,50 @@ void mouse_handler::show_attack_options(const unit_map::const_iterator& u)
...
+        }
+    }
+
+	// for optimalization, I think iterating over ranges here, would be better
+
+	int min = *attackable_distances.begin();
+	int n = *std::prev(attackable_distances.end());
+	const map_location& a = u->get_location();
+
+	int cz = a.y - (a.x - (a.x&1)) / 2;
```
Comment (CelticMinstrel) [inline src/mouse_events.cpp:1540]: As with the other PR, let's use `floor(a.x / 2.0)` here.

File: src/mouse_events.cpp [hunk 2]
```diff
@@ -1484,8 +1515,50 @@ void mouse_handler::show_attack_options(const unit_map::const_iterator& u)
...
+	for (int dx = -n; dx <= n; ++dx) {
+		for (int dy = std::max(-n, -dx-n); dy <= std::min(n, -dx+n); ++dy) {
+			int dz = -dx - dy;
+			if (std::abs(dx) + std::abs(dy) + std::abs(dz) < 2*min) {
+				continue;
+			}
+			int z = cz + dz;
+			int x = a.x + dx;
+			// convert back to offset
+			tiles.emplace_back(x, z + (x - (x&1)) / 2);
```
Comment (CelticMinstrel) [inline src/mouse_events.cpp:1552]: `floor(x / 2.0)`

File: src/mouse_events.cpp (hunk 1 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:1540]: This is generally used code man, this is the current best way.
...
Do we really have to do this?

Comment (CelticMinstrel) [reply src/mouse_events.cpp:1540]: I think it's clearer if code expresses is intent rather than using obscure bitwise operations. Yes, even if "everyone else" uses obscure bitwise operations.

Based on the link you put in the code a few lines up, there's the suggestion of using `abs(x)%2` instead of `x&1`, which would be slightly better, but `(x - x%2) / 2` still looks weird (and divides _twice_!), so I think the floor is still preferred.

Although, it would be even better if the logic is extracted out into a function, probably a member function of `map_location`. It could be called `to_cubic` or similar and return a `tuple<int,int,int>` or a new custom struct with three members `q, r, s`.

Comment (hladkyjakub) [reply src/mouse_events.cpp]: Bruh

File: src/mouse_events.cpp (hunk 2 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp]: Here you go

File: src/mouse_events.cpp [hunk 3]
```diff
@@ -1476,16 +1507,58 @@ std::set<map_location> mouse_handler::get_adj_enemies(const map_location& loc, i
...
+    for (const auto& attack : attacks) {
+        for (int i = attack.min_range(); i <= attack.max_range(); ++i) {
+            attackable_distances.insert(i);
+        }
+    }
+
+	// for optimalization, I think iterating over ranges here, would be better
+
+	int min = *attackable_distances.begin();
+	int n = *std::prev(attackable_distances.end());
```
Comment (soliton-) [inline src/mouse_events.cpp:1537]: Why are these dereferences allowed? I suggest checking for empty.

Comment (hladkyjakub) [reply src/mouse_events.cpp:1537]: What do you mean?

Comment (hladkyjakub) [reply src/mouse_events.cpp:1537]: I fo check if the attacks are empty already

Comment (CelticMinstrel) [reply src/mouse_events.cpp:1537]: The set is known to be non-empty because the only way for it to be empty is if there are no attacks, and if there are no attacks then it doesn't even reach this line due to the check on line 1510.

Both dereferences are guaranteed to be valid if there's at least one element, and we know there's at least one element. (I suppose you could add `assert(!attackable_distances.empty());` above these lines to make it clearer that we know there's at least one element.)

Comment (hladkyjakub) [reply src/mouse_events.cpp:1537]: Ok

**Comment by CelticMinstrel** [comment]
Okay, so after some investigation and testing, I've determined that we _already have_ a function that does the conversion you want. It doesn't match up very well with the code on redblobgames.com, and it's probably a lot less efficient, but it appears to produce equivalent results.

The function is `map_location::get_in_basis_N_NE`. It returns a pair where `first` is what redblobgames.com calls `s`, and `second` is what redblobgames.com calls `q`.  That means that if you need what redblobgames calls `r` you can calculate it as `-second-first`.

So, please use `get_in_basis_N_NE` instead of the `floor` formula I previously asked you to use. If you need help figuring out exactly how to use `get_in_bases_N_NE` to find the same answer, feel free to ask.

**Comment by hladkyjakub** [comment]
Why? Why? Why?
If it's less efficient, why to use it?
Btw also found out how to check for big hexagon, not rectangle so no need for distance_between check, so it's more effective, also works well for max-min ranges, I thought, I could implement that.

File: src/mouse_events.cpp (hunk 3 above)
Comment (soliton-) [reply src/mouse_events.cpp:1537]: Unless there is code elsewhere that checks min and max range there is no reason to expect the set to be non-empty. I suggest to not add an assert.

**Comment by soliton-** [comment]
> Btw also found out how to check for big hexagon, not rectangle so no need for distance_between check, so it's more effective, also works well for max-min ranges, I thought, I could implement that.

Implement that in a separate function.

File: src/mouse_events.cpp (hunk 3 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:1537]: I added assert, if the set is empty already.
You've seen that?

**Comment by hladkyjakub** [comment]
> Implement that in a separate function. `map_location` might be a good place for it.

Ok, I can do that.

File: src/mouse_events.cpp (hunk 3 above)
Comment (soliton-) [reply src/mouse_events.cpp:1537]: Yes, hence now it will abort the process instead of more reasonable behavior.

Comment (hladkyjakub) [reply src/mouse_events.cpp:1537]: It won't reach that assert.

Comment (soliton-) [reply src/mouse_events.cpp:1537]: No clue how you come to that conclusion. Where is the code that guarantees it?

**Comment by soliton-** [comment]
A function that returns a list of locations in a certain distance range to a center location sounds like something that could be useful to expose to Lua and WFL as well.

File: src/mouse_events.cpp (hunk 3 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:1537]: > No clue how you come to that conclusion. Where is the code that guarantees it?

Tell me example, it could fail.

Comment (soliton-) [reply src/mouse_events.cpp:1537]: Pick min and max ranges so nothing is added to the set.

Comment (hladkyjakub) [reply src/mouse_events.cpp:1537]: Isn't that true only for invalid min and max ranges?

**Comment by CelticMinstrel** [comment]
> Why? Why? Why?
> If it's less efficient, why to use it?

Because it makes clear your intent. If it's too inefficient, we can just make it more efficient, but the point is that we want to be able to look at the code and know what it's doing.

> A function that returns a list of locations in a certain distance range to a center location sounds like something that could be useful to expose to Lua and WFL as well.

Sounds like `wesnoth.map.get_hexes_in_radius` (or in C++, the global function `get_tiles_in_radius`). Please don't reimplement things that we already have, just use them!

If you're looking for _just_ the single ring at exactly the target radius, that also exists too (though only in C++) – it's called `get_tile_ring`.

Both functions are in `"pathutils.hpp"`.

**Comment by CelticMinstrel** [comment]
> The function is `map_location::get_in_basis_N_NE`. It returns a pair where `first` is what redblobgames.com calls `s`, and `second` is what redblobgames.com calls `q`. That means that if you need what redblobgames calls `r` you can calculate it as `-second-first`.

Looking more closely at the info on redblobgames.com, it seems I described this wrong. On redblobgames.com, they use `q` for the north-south axis with south positive; `r` for the northwest-southeast axis with southeast positive; and `s` for the northeast-southwest axis with southwest positive.

So that means that our `first` (north-south axis, north positive) is actually their `-q`, and our `second` (northeast-southwest axis, northeast positive) is actually their `-s`. By extension, their `r` would be `first+second` in our system.

**Comment by CelticMinstrel** [comment]
> The function is `map_location::get_in_basis_N_NE`. It returns a pair where `first` is what redblobgames.com calls `s`, and `second` is what redblobgames.com calls `q`. That means that if you need what redblobgames calls `r` you can calculate it as `-second-first`.

Looking more closely at the info on redblobgames.com, it seems I misunderstood something about how the axial/cube coordinates work. I'll get back to you on this later, but probably don't make the changes yet.

**Comment by soliton-** [comment]
Obviously if the function already exists don't add another one. I did not check if any exists at all.

File: src/mouse_events.cpp (hunk 3 above)
Comment (soliton-) [reply src/mouse_events.cpp:1537]: Can you elaborate on what that is and where that is handled?

Comment (hladkyjakub) [reply src/mouse_events.cpp:1537]: > Can you elaborate on what that is and where that is handled?

I'm sure it happen only when min_range > max_range,
Which is invalid. ...
But I see you point, I'll add a fallback if someone sets those numbers wrongly.

Comment (hladkyjakub) [reply src/mouse_events.cpp:1537]: *instead of the assert.

Comment (hladkyjakub) [reply src/mouse_events.cpp:1539]: (Same for the othor UI PR)

**Comment by hladkyjakub** [comment]
okay... I'm waiting.

**Comment by CelticMinstrel** [comment]
What I missed in my earlier perusal is that the axial coordinates described on redblobgames.com use axes that are _not_ aligned with the directions of the hex grid. (They are in fact the "diagonal directions" of the hex grid.) The basis_N_NE coordinates used in Wesnoth for rotations are _not_ like that – their axes _are_ aligned with the directions of the hex grid. So it's doubtful whether they are interchangeable like I implied earlier.

So, this basically leaves you with two options:

1. You can try to figure out a way to use `get_in_basis_N_NE` to calculate whatever you need to calculate. I think you were looking for distance, so the N-NE basis may be satisfactory, but I'm not sure.
2. You can create a new function in `map_location` which converts to cube or axial coordinates by redblobgames's definition. I think it could be called `get_in_basis_E_SW` to mirror the name of the existing function, but I'm open to other naming suggestions. It should probably return a struct with elements `q` and `r` (which I think you've been calling `x` and `y` in the code so far), and maybe `s` either as an element or a member function.

The main point here is that putting the logic behind a function makes it clearer what you're doing. I'd be less opposed to using the bit-manipulation method if it's in a function.

**Comment by hladkyjakub** [comment]
>     2. You can create a new function in `map_location` which converts to cube or axial coordinates by redblobgames's definition. I think it could be called `get_in_basis_E_SW` to mirror the name of the existing function, but I'm open to other naming suggestions. It should probably return a struct with elements `q` and `r` (which I think you've been calling `x` and `y` in the code so far), and maybe `s` either as an element or a member function.
> 
> The main point here is that putting the logic behind a function makes it clearer what you're doing. I'd be less opposed to using the bit-manipulation method if it's in a function.

I could try the 2nd option, you gave me, haven't ever created a function, included in the header file, is it complicated?
It's late today, I'll get to it tomorrow.

**Review by soliton- — CHANGES_REQUESTED** [review]
Just to be clear that I do not think this is ready to merge (neither was #9255 IMO).

The hex math should be put in a separate function (or use suitable existing functions) and documented properly (same as for #9255).

**Comment by hladkyjakub** [comment]
I fully agree, but got no time nor will now to be honest.

**Comment by CelticMinstrel** [comment]
Don't worry, I made the function for you. Once #9340 is merged you can rebase this and use the new function.

**Comment by hladkyjakub** [comment]
> Don't worry, I made the function for you. Once #9340 is merged you can rebase this and use the new function.

May you remind me, what part of code is problematic for you?

**Comment by hladkyjakub** [comment]
Nvm, found that

**Comment by hladkyjakub** [comment]
Someone any idea what I am supposed to do?

**Comment by Pentarctagon** [comment]
@CelticMinstrel  What function were you referring to?

**Comment by hladkyjakub** [comment]
@CelticMinstrel 
Why this does return an error?
to_cubic returns cubic_location, for sure
```
const map_location& a = u->get_location();
cubic_location center = a.to_cubic();
```
error: ‘cubic_location’ was not declared in this scope; did you mean ‘map_location’?

**Comment by hladkyjakub** [comment]
Ok, I'm back on the track.
@soliton- 
Would code like this:
```
cubic_location center = a.to_cubic();

	// Enumerate range in cubic coordinates
	for (int dx = -n; dx <= n; ++dx) {
		for (int dy = std::max(-n, -dx-n); dy <= std::min(n, -dx+n); ++dy) {
			int dz = -dx - dy;

			// Skip positions that are too close to center
			if (std::abs(dx) + std::abs(dy) + std::abs(dz) < 2*min) {
				continue;
			}

			// Create new cubic location
			cubic_location neighbor{
				center.q + dx,
				center.r + dy,
				center.s + dz
			};

			// Convert back to map coordinates and add to tiles
			tiles.push_back(map_location::from_cubic(neighbor));
		}
	}
	```
	resolve your issue?

**Comment by soliton-** [comment]
That looks like it should be a function inrange or so that returns or adds to a list of locations.

I don't know what min is but comparing to 2*min looks odd.

**Comment by hladkyjakub** [comment]
> That looks like it should be a function inrange or so that returns or adds to a list of locations.

The whole code, I posted should be a function defined in mouse_events.cpp?

> I don't know what min is but comparing to 2*min looks odd.

It has purpose, this is the old function, just rewritten with from/to_cubic.

**Comment by soliton-** [comment]
The function would be somewhere fitting. Probably near `cubic_location`.

**Comment by soliton-** [comment]
The function would be somewhere fitting. Probably near `cubic_location`.

**Comment by hladkyjakub** [comment]
> The function would be somewhere fitting. Probably near `cubic_location`.

I can do that.
Takes location and 2 numbers and returns vector of locations that are at least n1 and most n2 distant from the center, unchecked if existing
How to call it tho?

**Comment by hladkyjakub** [comment]
My ideas would be:
get_location_ring/ just get_ring or
locations_within_distance_range

**Comment by hladkyjakub** [comment]
@soliton- Would that work?

**Comment by soliton-** [comment]
Sure

**Comment by hladkyjakub** [comment]
Cool

**Comment by hladkyjakub** [comment]
After I'll have the basics done, what is the way to handle move&attack command UI?
I would put both ranged and melee weapons into the very same attack dialog.
This is very important as whaever I'll do, I belive shouldn't affect any content that doesn't use real ranged attacks, and as well shouldn't limit UI for units who have a real-ramged attack, but want to perform move&close-attack action

**Comment by hladkyjakub** [comment]
So the function is ready and working... but now I have to know what to do next

**Comment by hladkyjakub** [comment]
@soliton- @CelticMinstrel - You're the two who were involved in this UI,
(And I assume if nobody else reacts, and we three agree on something, then later nobody can come and say that he dislikes it and should be reworked from scratch)
What should the UI behave like?

**Comment by hladkyjakub** [comment]
@soliton- 
(And what to highlight when you have an archer selected and hover over an enemy?)
I believe whatever I'll do, I belive shouldn't affect any content that doesn't use real ranged attacks, and as well shouldn't limit UI for units who have a real-ranged attack, but want to perform move&close-attack action.

I suggest placing both ranged and melee weapons into the very same attack dialog, so by selecting your unit and then enemy, one attack dialog would pop up, giving you chance to either attack ranged, or to move&attack.

**Comment by ProditorMagnus** [comment]
When attack dialog is used, then unit moves exactly as close to the target as specific attack needs. When player desires different positioning, he can move before using dialog.

**Comment by hladkyjakub** [comment]
> When attack dialog is used, then unit moves exactly as close to the target as specific attack needs. When player desires different positioning, he can move before using dialog.

I like this approach, only I would be fan of: moves as little as possible to attack.
That makes more sence if you have archer with range 2-3 and enemy like 5 hexes away he moves only 2 tiles and then attacks.

**Comment by ProditorMagnus** [comment]
That is what I said. Before each movement, check if it is possible to attack, if not, do next movement.

**Comment by hladkyjakub** [comment]
> That is what I said. Before each movement, check if it is possible to attack, if not, do next movement.

That won't work as I need to know the attacking location forehand, so I can use it in the attack, it will be precauculated (easier to do too).
I would then pass one location for every weapon to the attack dialog, so it can count with it.

**Comment by hladkyjakub** [comment]
Ok, so the current system doesn't highlight attackable enemies, only changes border highlight on adjacent tiles reachable by the attacker
... that's new, huh

(on early 1.19.3, when I worked on it, the highlight was simply adding enemies to the reachmap, thus highlighting them correctly, but possible target enemies aren't highlighted currently, I ask what to do)

...
So how am I supposed to highlight the possible targets for a ranged attack in the new GUI?

**Comment by soliton-** [comment]
> So how am I supposed to highlight the possible targets for a ranged attack in the new GUI?

Where does this come from? Who or what is supposing you do something or other?

If you think some highlight is needed then surely you have some idea what that should look like?

I don't see anything special in your old screenshots either though so I don't know why you must do something special now.

**Comment by hladkyjakub** [comment]
> I don't see anything special in your old screenshots either though so I don't know why you must do something special now.
It was added after the screenshot, the same way the original PR did.
Ok, it certainly isn't a thing that would require immediate attention.

**Comment by hladkyjakub** [comment]
https://github.com/wesnoth/wesnoth/blob/master/src/mouse_events.cpp#L1528
dlg.show() keeps returning me false (and showing no dialog).
Where is the function defined? (I couldn't find anything in src/gui/dialogs)
Is the function somehow edited, but not defined for the gui2::dialogs::unit_attack class?

**Comment by hladkyjakub** [comment]
Ok, it shows that I canged 9 files in total which I definitely didn't touched, so probably pulling changes f*cked up somehow, or those are some kind of tmp files... no idea

**Comment by hladkyjakub** [comment]
... nvm, I don't see it now

**Comment by hladkyjakub** [comment]
There's code, I forgot to erase from some files, before mering, but have you got any idea why does it keep showing me that I've changed 42 files in src/modules/lua?

**Comment by ProditorMagnus** [comment]
You pushed some random commits to this branch. Remove them, rebase on master.

File: src/mouse_events.cpp [hunk 4]
```diff
@@ -749,47 +751,76 @@ map_location mouse_handler::current_unit_attacks_from(const map_location& loc) c
...
+		for (const auto& attack : attacks) {
+			for (int i = attack.min_range(); i <= attack.max_range(); ++i) {
+				attackable_distances.insert(i);
+			}
+		}
 	}
 
-	const map_location::direction preferred = loc.get_relative_dir(previous_hex_);
-	const map_location::direction second_preferred = loc.get_relative_dir(previous_free_hex_);
+	if(*attackable_distances.rbegin() <= 1){
```
Comment (soliton-) [inline src/mouse_events.cpp:763]: Instead of putting the rest of the function in the if invert the check and return early. You're also missing a check for empty before you assume there is one element.

Comment (soliton-) [reply src/mouse_events.cpp:763]: Well, I guess this is not the rest of the function. The diff is unreadable so I'll wait for the explanation.

Comment (hladkyjakub) [reply src/mouse_events.cpp:763]: @soliton- So what do you think?

Comment (hladkyjakub) [reply src/mouse_events.cpp:763]: @soliton- I don't mean to be rude, but I'm really waiting for your response, so I can move on with the work.

File: src/mouse_events.cpp [hunk 5]
```diff
@@ -749,14 +751,41 @@ map_location mouse_handler::current_unit_attacks_from(const map_location& loc) c
...
+		for (const auto& attack : attacks) {
+			for (int i = attack.min_range(); i <= attack.max_range(); ++i) {
+				attackable_distances.insert(i);
+			}
+		}
 	}
 
+	map_location res;
+
+	if(attackable_distances.empty()) {return res} 	//invalid attack ranges
```
Comment (Vultraz) [inline src/mouse_events.cpp:765]: Cleanup suggestions:
```cpp
	if(attackable_distances.empty()) {
		return map_location{};
	}

	// ranged attack
	if(attackable_distances.back() > 1) {
		if(utils::contains(attackable_distances, distance_between(selected_hex_, loc)) {
			return selected_hex_;
		}

		int best_move = -1;
		map_location res;

		for(const pathfind::paths::step& step : current_paths_.destinations) {
			const map_location& dst = step.curr;

			if(utils::contains(attackable_distances, distance_between(loc, dst)) {
				if(step.move_left > best_move) {
					best_move = step.move_left;
					res = dst;
				}
			}
		}

		return res;
	}
```

Comment (hladkyjakub) [reply src/mouse_events.cpp]: ok

File: src/mouse_events.cpp [hunk 6]
```diff
@@ -926,7 +964,7 @@ void mouse_handler::move_action(bool browse)
 
 		src = selected_hex_;
 		orig_paths = current_paths_;
-		attack_from = current_unit_attacks_from(hex);
+		attack_from = current_unit_attacks_from(hex); //Where does the attack dialog gonna be showed from, only one tile now
```
Comment (soliton-) [inline src/mouse_events.cpp:967]: Is this a question or a clarifying comment or what?

File: src/mouse_events.cpp [hunk 7]
```diff
@@ -1525,7 +1565,7 @@ int mouse_handler::show_attack_dialog(const map_location& attacker_loc, const ma
 	// bc_widget_data_vector won't be empty when it reaches here.
 	gui2::dialogs::unit_attack dlg(attacker, defender, std::move(bc_vector), best, bc_widget_data_vector, leadership_bonus);
 
-	if(dlg.show()) {
+	if(dlg.show()) { //dlg.show() is what actually shows the dialog
```
Comment (soliton-) [inline src/mouse_events.cpp:1568]: I'm all for comments explaining non-obvious code but this seems a bit much.

File: src/mouse_events.cpp (hunk 6 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:967]: It's supposed to be a clarifying comment ,:D

File: src/mouse_events.cpp (hunk 7 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:1568]: Okay, it just took me half an hour to find out that what actually shows the dialog is in an if... but yes, seems obviois now

File: src/mouse_events.cpp (hunk 6 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:967]: What do you want to be done about it?

File: src/mouse_events.cpp (hunk 7 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:1568]: removing it

File: src/mouse_events.cpp [hunk 8]
```diff
@@ -1630,6 +1670,69 @@ std::set<map_location> mouse_handler::get_adj_enemies(const map_location& loc, i
 	return res;
 }
 
+/**
+ * NOT NEEDED now, keeping for future work on ranged attacks UI
```
Comment (soliton-) [inline src/mouse_events.cpp:1674]: This does not seem sensible. If you want to keep code around for future work then do that privately. No point in leaving in dead code.

File: src/mouse_events.cpp (hunk 6 above)
Comment (soliton-) [reply src/mouse_events.cpp:967]: Should be written to be understandable.

Comment (hladkyjakub) [reply src/mouse_events.cpp:967]: Would:
//the attack dialog is showed like if the unit is standing at the hex, this may be not the best approach
Work?

File: src/mouse_events.cpp (hunk 4 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp]: resolved ig

**Comment by hladkyjakub** [comment]
So ... can it be merged now?

File: src/mouse_events.cpp (hunk 6 above)
Comment (soliton-) [reply src/mouse_events.cpp]: `// the attack dialog is shown as if the unit is standing at the hex, this may be not the best approach`

What hex? The selected one, I guess? And this is talking about real ranged attacks or what? Why is this comment here and not inside the function?

**Review by soliton- — APPROVED** [review]
Disregarding the confusing comment this looks good to me.

**Comment by soliton-** [comment]
> > Disregarding the confusing comment this looks good to me.
> 
> I'd like to just mention that in there, if you want, you can suggest how should it be.
> 
> So ... what happens now?

I've asked multiple questions to understand what the comment is meant to express. You could answer them. Otherwise I cannot suggest anything since I have no idea what you're trying to say.

**Comment by hladkyjakub** [comment]
> I've asked multiple questions to understand what the comment is meant to express. You could answer them. Otherwise I cannot suggest anything since I have no idea what you're trying to say.

Sorry, as I marked it as resolved, I couldn't see it.
The hex I mean (the variable named hex)
Maybe I should mention that this isn't ideal _because_ of ranged attacks?
Maybe I should move it to the function.

**Comment by soliton-** [comment]
> Maybe I should mention that this isn't ideal _because_ of ranged attacks? Maybe I should move it to the function.

That would make sense to me.

**Comment by hladkyjakub** [comment]
Ok

**Comment by soliton-** [comment]
Since it seems difficult to get to an understandable comment I'll merge this without the comment in the next days.

**Comment by hladkyjakub** [comment]
> Since it seems difficult to get to an understandable comment I'll merge this without the comment in the next days.

Cool.
Can you rephrase it then?

**Comment by soliton-** [comment]
I've already rephrased it and that was ignored. I'm not really interested in wasting more time with this.

**Comment by hladkyjakub** [comment]
> I've already rephrased it and that was ignored. I'm not really interested in wasting more time with this.

I'm sorry, but I've searched throught this whole PR and have found no rephrasing suggestion. You may haven't actually posted it.

**Comment by hladkyjakub** [comment]
> https://github.com/wesnoth/wesnoth/pull/9286#discussion_r2004405965

Yes, after that I came with two new versions, following your advise... and I don't know what's wrong with them. (And sorry for being pointlich, but you haven't offered a rephrase)

**Comment by soliton-** [comment]
This is getting ridiculous. Is your browser broken or something? I linked to the last comment in that thread there is no follow up message from you and the next commit uses the same broken english that I corrected there.

I have no idea where you offered two new versions.

**Comment by ProditorMagnus** [comment]
I expect ZK thought monospace meant quote and skipped reading it.

**Comment by hladkyjakub** [comment]
Edited to this then... hope that'll work
// the attack dialog is shown as if the unit is standing at the hex, this may be not the best approach

**Comment by hladkyjakub** [comment]
@soliton- Is there a problem? I'd like to get it merged...

File: src/mouse_events.cpp [hunk 9]
```diff
@@ -749,8 +752,44 @@ map_location mouse_handler::current_unit_attacks_from(const map_location& loc) c
 		if(!target_eligible) {
 			return map_location();
 		}
+
+		const auto& attacks = source_unit->attacks();
+		for (const auto& attack : attacks) {
```
Comment (Vultraz) [inline src/mouse_events.cpp:757]:
```suggestion
		for(const auto& attack :  source_unit->attacks()) {
```

File: src/mouse_events.cpp [hunk 10]
```diff
@@ -749,8 +752,44 @@ map_location mouse_handler::current_unit_attacks_from(const map_location& loc) c
...
+			for (int i = attack.min_range(); i <= attack.max_range(); ++i) {
+				attackable_distances.insert(i);
+			}
+		}
+	}
+
+
```
Comment (Vultraz) [inline src/mouse_events.cpp:764]: Please remove excess newlines.

File: src/mouse_events.cpp [hunk 11]
```diff
@@ -749,8 +752,44 @@ map_location mouse_handler::current_unit_attacks_from(const map_location& loc) c
...
+	//invalid attack ranges
+	if(attackable_distances.empty()) {
+		return map_location{};
+	}
+
+	//ranged attack
+	if(*attackable_distances.rbegin() > 1){
+		int distance = distance_between(selected_hex_, loc);
+		if (attackable_distances.find(distance) != attackable_distances.end() ) {
```
Comment (Vultraz) [inline src/mouse_events.cpp:774]: Does `utils::contains` not work here (and below)?

File: src/mouse_events.cpp [hunk 12]
```diff
@@ -788,8 +827,8 @@ map_location mouse_handler::current_unit_attacks_from(const map_location& loc) c
 			}
 		}
 	}
-
 	return res;
+
```
Comment (Vultraz) [inline src/mouse_events.cpp:831]: Please undo these two whitespace changes.

File: src/mouse_events.cpp [hunk 13]
```diff
@@ -1354,7 +1393,8 @@ int mouse_handler::show_attack_dialog(const map_location& attacker_loc, const ma
 		defender = board.units().find(defender_loc);
 
 		if(!attacker || !defender) {
-			ERR_NG << "One fighter is missing, can't attack";
+			if (!attacker) {ERR_NG << "Attacker is missing, can't attack";}
```
Comment (Vultraz) [inline src/mouse_events.cpp:1396]: Why is this additional logging needed?

File: src/mouse_events.cpp (hunk 9 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:757]: Why not attacks?

File: src/mouse_events.cpp (hunk 10 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:764]: Ok

File: src/mouse_events.cpp (hunk 12 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:831]: Ok... you commented only one

File: src/mouse_events.cpp (hunk 13 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:1396]: More telling to you, it isn't _needed_.

File: src/mouse_events.cpp (hunk 11 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:774]: Rephrase the question, please.

File: src/mouse_events.cpp (hunk 12 above)
Comment (Vultraz) [reply src/mouse_events.cpp:831]: You added a line and removed a line. Just move `return res` back to the end of the function.

File: src/mouse_events.cpp (hunk 9 above)
Comment (Vultraz) [reply src/mouse_events.cpp:757]: You don't need the separate local variable `attacks` (it can be removed). Just loop over `source_unit->attacks()` directly.

File: src/mouse_events.cpp (hunk 12 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:831]: Ok ^^

File: src/mouse_events.cpp (hunk 11 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:774]: I can try.
What would be the advantage of the method?

Comment (Vultraz) [reply src/mouse_events.cpp:774]: Cleaner, easier-to-read code, and we use `utils::contains` widely in the codebase.

File: src/mouse_events.cpp (hunk 9 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:757]: yes, that has no point in there

File: src/mouse_events.cpp (hunk 11 above)
Comment (hladkyjakub) [reply src/mouse_events.cpp:774]: Works fine :D

**Review by Vultraz — APPROVED** [review]

**Comment by hladkyjakub** [comment]
Nice.
If someone would be interested in trying out how the basic UI works, 
you can download _Ranged Attacks Preshow_ add-on on 1.19 add-on server (use latest build).

Also,...  I want to continue work on ranged attacks UI as many aspects of it need to be improved.
What should I do/where should I do it? (should I open a discussion thread, should I continue here?)

For the further development:

1. Should there be further development of the functionality required (the non-UI part)?

2. How can we make projectile animations to travel from attackers to defenders' hexes?

3. Should a way to connect combatants with the red arrow be found? Or should a better graphic for that be made?

4. New 1.19 UI doesn't allow to highlight all possible targets, hat way should they be highlighted?
